py4j-python/src/py4j/java_gateway.py

```diff
@@ -1182,12 +1184,14 @@ def launch_gateway(cls, port=0, jarpath="", classpath="", javaopts=[],
         :param die_on_exit: if `True`, the Java gateway process will die when
             this Python process exits or is killed.
 
+        Additional keyword arguments are passed along to the newly created
+        instance.
+
         :rtype: a :class:`JavaGateway <py4j.java_gateway.JavaGateway>`
             connected to the `Gateway` server.
         """
         _port = launch_gateway(port, jarpath, classpath, javaopts, die_on_exit)
-        gateway = JavaGateway(GatewayClient(port=_port))
-        return gateway
+        return cls(GatewayClient(port=_port), **kwargs)
```

Comment:
nice catch (cls instead of JavaGateway).

The main issue I see is that GatewayClient is deprecated and should be passed in an instance of GatewayParameters. Maybe instead of kwargs, the method should access an instance of GatewayParameters and CallbackServerParameters (these could be extended to pass extra parameters to a subclass of Gateway).